src/binaryContentRenderer.ts

```diff
@@ -0,0 +1,116 @@
+    private static contentLengthPerTableCell: number = BinaryContentRenderer.hexNumberPrefixLength + BinaryContentRenderer.hexNumberLength + BinaryContentRenderer.markdownTableDelimiterLength;
+    private static binaryDataMarkdownHeader: string = localize("embeddedContent.tableHeader", "|Offset|0|1|2|3|4|5|6|7|\r\n|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|");
+
+    /**
+     * Creates an instance of the binary content renderer.
+     * @param content The string contents to be rendered as markdown.
+     */
+    private constructor(private readonly content: string) {
+    }
+
+    /**
+     * Attempts to create an instance of the binary content renderer based on a SARIF log, run index and artifact index
+     * Returns undefined if the artifact contents cannot be found, or the content is not binary content.
+     * @param log The SARIF log.
+     * @param runIndex The run index.
+     * @param artifactIndex The artifact index.
+     */
+    public static tryCreateFromLog(log: sarif.Log, runIndex: number, artifactIndex: number): BinaryContentRenderer | undefined {
+        const run: sarif.Run | undefined = log.runs[runIndex];
+        if (!run) {
+            return undefined;
+        }
+
+        if (!run.artifacts) {
+            return undefined;
+        }
+
+        const artifact: sarif.Artifact = run.artifacts[artifactIndex];
+        if (!artifact) {
+            return undefined;
+        }
+
+        if (!artifact.location) {
+            return undefined;
+        }
+
+        if (!artifact.contents) {
+            return undefined;
+        }
+
+        if (!artifact.contents.binary) {
+            return undefined;
+        }
+
+        return new BinaryContentRenderer(artifact.contents.binary);
+    }
+
+    public renderAsMarkdown(displayFileName: string): string {
+        const contentBuffer: Buffer = Buffer.from(this.content, 'base64');
+        let markDownContent: string = localize("embeddedContent.fileInfoHeader", "# File {0}\r\n", displayFileName);
+        markDownContent = markDownContent.concat(localize("embeddedContent.fileInfoHeader", "Total bytes {0}\r\n", contentBuffer.length));
+        markDownContent =  markDownContent.concat(BinaryContentRenderer.binaryDataMarkdownHeader);
+        for (let bufferIndex: number = 0; bufferIndex < contentBuffer.length; bufferIndex++) {
+            const bufferByte: number = contentBuffer[bufferIndex];
+            // When we hit the start of a new row, add the offset marker.
+            if (bufferIndex % BinaryContentRenderer.bytesPerRow === 0) {
+                markDownContent = markDownContent.concat(`\r\n|0x${bufferIndex < 16 ? '0' : ''}${bufferIndex.toString(16)}`);
+            }
+
+            // Add |0xFF, or |0xFF|\r\n depending on if we are at the end of a row.
+            markDownContent = markDownContent.concat(`|0x${bufferByte < 16 ? '0' : ''}${bufferByte.toString(16)}${((bufferIndex + 1) % BinaryContentRenderer.bytesPerRow === 0) ? '|\r\n' : ''}`);
```

Comment:
>\r\n [](start = 180, length = 4)

is this extension supposed to be cross platform? I would think so...

Reply:
Yep. I'll see if I can find the correct characters.

---
In reply to: [410424869](https://github.com/microsoft/sarif-vscode-extension/pull/252#discussion_r410424869) [](ancestors = 410424869)